rclcpp/test/rclcpp/test_utilities.cpp

```diff
@@ -78,3 +83,130 @@ TEST(TestUtilities, multi_init) {
+    EXPECT_THROW(
+      {interrupt_guard_condition = context1->get_interrupt_guard_condition(&wait_set);},
+      rclcpp::exceptions::RCLError);
+  }
+
+  {
+    interrupt_guard_condition = context1->get_interrupt_guard_condition(&wait_set);
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rclcpp", rcl_guard_condition_fini, RCL_RET_ERROR);
+    EXPECT_THROW(
+      {context1->release_interrupt_guard_condition(&wait_set);},
+      rclcpp::exceptions::RCLError);
+  }
+
+  {
+    interrupt_guard_condition = context1->get_interrupt_guard_condition(&wait_set);
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rclcpp", rcl_guard_condition_fini, RCL_RET_ERROR);
+    EXPECT_NO_THROW({context1->release_interrupt_guard_condition(&wait_set, std::nothrow);});
+  }
+
+  {
+    EXPECT_THROW(
+      context1->release_interrupt_guard_condition(nullptr),
+      std::runtime_error);
+  }
+
+  // Test it works after restore mocks
+  interrupt_guard_condition = context1->get_interrupt_guard_condition(&wait_set);
+  EXPECT_NE(nullptr, interrupt_guard_condition);
+  EXPECT_NO_THROW(context1->release_interrupt_guard_condition(&wait_set));
+
+  rclcpp::shutdown(context1);
+}
+
+
+TEST(TestUtilities, test_context_init_shutdown_fails) {
+  auto context = std::make_shared<rclcpp::contexts::DefaultContext>();
+
+  {
+    auto context_fail_init = std::make_shared<rclcpp::contexts::DefaultContext>();
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_init, RCL_RET_ERROR);
+    EXPECT_THROW(context_fail_init->init(0, nullptr), rclcpp::exceptions::RCLError);
+  }
+
+  {
+    auto context_fail_init = std::make_shared<rclcpp::contexts::DefaultContext>();
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_logging_configure_with_output_handler, RCL_RET_ERROR);
+    EXPECT_THROW(context_fail_init->init(0, nullptr), rclcpp::exceptions::RCLError);
+  }
+
+  {
+    context->init(0, nullptr);
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rclcpp", rcl_trigger_guard_condition, RCL_RET_ERROR);
+    // This will log a message, no throw expected
+    EXPECT_NO_THROW(context->shutdown(""));
+  }
+
+  {
+    context->init(0, nullptr);
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rclcpp", rcl_shutdown, RCL_RET_ERROR);
+    EXPECT_THROW(context->shutdown(""), rclcpp::exceptions::RCLError);
+  }
+
+  {
+    context->init(0, nullptr);
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rclcpp", rcl_logging_fini, RCL_RET_ERROR);
+    // This will log a message, no throw expected
+    EXPECT_NO_THROW(context->shutdown(""));
+  }
+
+  {
+    auto context_to_destroy = std::make_shared<rclcpp::contexts::DefaultContext>();
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rclcpp", rcl_trigger_guard_condition, RCL_RET_ERROR);
+    // This will log a message, no throw expected
+    EXPECT_NO_THROW(
+    {
+      // context_to_destroy.~rclcpp::Context();
```

Comment:
@brawner I had that part of the code working, but it is not working once I add the expect_no_throw, I think it's similar to the code you had in the wait_policies PR

Reply:
If you want to call the constructor, you can call `reset()` on the shared_ptr (as long as its use count is 1).
```
EXPECT_NO_THROW(context_to_destroy.reset());
```

Reply:
Thanks! eb70a4b